File: backend/src/main/java/com/readingbuddy/backend/domain/dashboard/dto/response/StageMasteryResponse.java
package com.readingbuddy.backend.domain.dashboard.dto.response;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;
import java.util.List;

@Getter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class StageMasteryResponse {

    private String stage;
    private List<KcMastery> kcMasteries;
    private Double averageMastery;

    @Getter
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class KcMastery {
        private Long kcId;
        private String kcCategory;
        private Float pLearn;
        private Float pTrain;
        private Float pGuess;
        private Float pSlip;
        private LocalDateTime updatedAt;
    }
}

File: reading_buddy_app/lib/features/dashboard/data/models/stage_mastery_response.dart
import 'package:json_annotation/json_annotation.dart';

part 'stage_mastery_response.g.dart';

@JsonSerializable()
class StageMasteryResponse {
  final String? stage;
  final List<KcMastery>? kcMasteries;
  final double? averageMastery;

  StageMasteryResponse({
    this.stage,
    this.kcMasteries,
    this.averageMastery,
  });

  factory StageMasteryResponse.fromJson(Map<String, dynamic> json) =>
      _$StageMasteryResponseFromJson(json);

  Map<String, dynamic> toJson() => _$StageMasteryResponseToJson(this);

  /// 숙련도를 퍼센트로 변환 (0-100)
  double get masteryPercent => (averageMastery ?? 0.0) * 100.0;
}

@JsonSerializable()
class KcMastery {
  final int? kcId;
  final String? kcCategory;
  final double? pLearn;
  final double? pTrain;
  final double? pGuess;
  final double? pSlip;
  final DateTime? updatedAt;

  KcMastery({
    this.kcId,
    this.kcCategory,
    this.pLearn,
    this.pTrain,
    this.pGuess,
    this.pSlip,
    this.updatedAt,
  });

  factory KcMastery.fromJson(Map<String, dynamic> json) =>
      _$KcMasteryFromJson(json);

  Map<String, dynamic> toJson() => _$KcMasteryToJson(this);
}
